pkg/workflow/task/bootstrap/bootstrap.go

```diff
@@ -2,75 +2,146 @@ package bootstrap
-type OptionsRetriever interface {
-	BootstrapClusterOpts(*cluster.Spec) ([]bootstrapper.BootstrapClusterOption, error)
-}
-
-// Bootstrapper creates and destroys bootstrap clusters. It is satisfied by the bootstrap package
-// and exists predominently for testability.
-type Bootstrapper interface {
-	// CreateCluster creates a new local cluster. It does not contain any EKS-A components.
-	CreateBootstrapCluster(
-		context.Context,
-		*cluster.Spec,
-		...bootstrapper.BootstrapClusterOption,
-	) (*types.Cluster, error)
-
-	// DeleteBootstrapCluster deletes a local cluster created with CreateCluster.
-	DeleteBootstrapCluster(ctx context.Context, cluster *types.Cluster, isUpgrade bool) error
-}
-
 // CreateClusters creates a functional Kubernetes cluster that can be used to faciliate
 // EKS-A operations. The bootstrap cluster is populated in the context using
-// workflow.WithBootstrapCluster for subsequent tasks.
+// WithCluster for subsequent tasks.
 type CreateCluster struct {
 	// Spec is the spec to be used for bootstrapping the cluster.
 	Spec *cluster.Spec
 
 	// Options supplies bootstrap cluster creation options.
-	Options OptionsRetriever
+	Options ClientOptionsRetriever
 
 	// Bootstrapper is used to create the cluster.
-	Bootstrapper Bootstrapper
+	Client Client
+
+	// Writer writes files to a specific directory context.
+	Writer filewriter.FileWriter
 }
 
 // RunTask satisfies workflow.Task.
 func (t CreateCluster) RunTask(ctx context.Context) (context.Context, error) {
-	opts, err := t.Options.BootstrapClusterOpts(t.Spec)
+	clusterName := t.Spec.Cluster.Name
+
+	// Ensure the cluster doesn't already exist.
+	exists, err := t.Client.ClusterExists(ctx, clusterName)
 	if err != nil {
 		return ctx, err
 	}
+	if exists {
+		return ctx, fmt.Errorf("cluster %v already exists", clusterName)
+	}
 
-	cluster, err := t.Bootstrapper.CreateBootstrapCluster(ctx, t.Spec, opts...)
+	// Create the bootstrap cluster.
+	opts := t.Options.GetCreateBootstrapClusterOptions(t.Spec)
+	err = t.Client.CreateBootstrapCluster(ctx, clusterName, opts)
 	if err != nil {
 		return ctx, err
 	}
 
-	return WithCluster(ctx, cluster), nil
+	// Retrieve the kubeconfig and write it to non-volatile storage for later use.
+	kubeconfig, err := t.Client.GetKubeconfig(ctx, t.Spec.Cluster.Name)
+	if err != nil {
+		return ctx, err
+	}
+
+	kubeconfigPath, err := t.Writer.Write(toKubeconfigFilename(t.Spec.Cluster.Name), kubeconfig)
+	if err != nil {
+		return ctx, err
+	}
+
+	ctx = WithCluster(ctx, &types.Cluster{
+		Name:           t.Spec.Cluster.Name,
+		KubeconfigFile: kubeconfigPath,
+	})
+
+	return ctx, nil
 }
 
-// DeleteCluster deletes a bootstrap cluster. It expects the bootstrap cluster to be
-// populated in the context using workflow.WithBootstrapCluster.
+// KubernetesClient is used to interact with a Kubernetes cluster.
+type KubernetesClient interface {
+	Get(ctx context.Context, kubeconfig string, key corev1.ObjectReference, obj runtime.Object) error
```

Comment:
Although it makes sense from a concept perspective, I fear `ObjectReference` is not the best input for this kind of method. It contains things like UID, ResourceVersion and even FieldPath. We would need to document that those fields are not needed/ignored but still, it complicates the interface.

I would suggest to either have our own type with the minimum set of fields or maybe use a signature similar to https://github.com/aws/eks-anywhere/blob/main/pkg/clients/kubernetes/unauth.go#L41, which only requires name and namespace.

Reply:
Yea, it makes sense to descope it. I think there's a `NamespacedName` somewhere in apimachinery or some other Kubernetes lib?